### PR TITLE
Replace "path.__fspath__()" with "fspath(path)"

### LINKED ISSUE #419: In the async file code, use trio._util.fspath instead of calling __fspath__ directly

### PR DESCRIPTION
**Pull request by pmav99**
Fixes #419

### CODE REVIEW

File: trio/_path.py [hunk 1]
```diff
@@ -142,7 +142,7 @@ def __repr__(self):
 
     def __fspath__(self):
         try:
-            return self._wrapped.__fspath__()
+            return fspath(self._wrapped)
         # python3.5 compat
         except AttributeError:  # pragma: no cover
             return str(self)
```
Comment (njsmith) [inline trio/_path.py]: Tests are failing, and I think it's because this fallback isn't being hit anymore.

Comment (pmav99) [reply trio/_path.py]: You are right. I think the issue though is that the PEP implementation does not work with pre Python 3.6 pathlib objects.

I just pushed a new commit where I added that check on `fspath()` itself. If it works, then this workaround can probably be removed.

**Comment by njsmith** [comment]
Tests are still failing.

**Comment by pmav99** [comment]
@njsmith 
The tests are still failing on python 3.7, but I think that the failures are not related to the commit.

So the PR updates the `fspath()` implementation so that it works with Python 3.5 pathlib and removes the workaround from `trio.Path.__fspath__()`.

**Comment by njsmith** [comment]
Yep, looks good to me now. We'll have to wait for #434 to fix the CI stuff before merging though (since we have branch protection turned on).

**Comment by njsmith** [comment]
Cycling to trigger a new test run

**Comment by njsmith** [comment]
The complaint from codecov seems to be because `trio._util.fspath` doesn't have full coverage yet, and this PR touched it, so it's a bit cranky about the PR. I'm not bothered by this, so in it goes :-). (It would be good to get proper test coverage for our `fspath` implementation at some point, in case that's something you're interested in – should be pretty easy.)

Thanks @pmav99! And as per [our contributing docs](https://trio.readthedocs.io/en/latest/contributing.html#joining-the-team), watch out for a github invite.

**Comment by pmav99** [comment]
@njsmith 
The relevant tests from stdlib + one more test case for pathlib objects:
``` python
# -*- coding: utf-8 -*-

""" Backport of os.PathLike/os.fspath for Python < 3.6. """

# Upstream implementations:
# tests: https://github.com/python/cpython/blob/da6c3da6c33c6bf794f741e348b9c6d86cc43ec5/Lib/test/test_os.py#L3527-L3571
# fspath: https://github.com/python/cpython/blob/da6c3da6c33c6bf794f741e348b9c6d86cc43ec5/Lib/os.py#L1029-L1058
# PathLike: https://github.com/python/cpython/blob/da6c3da6c33c6bf794f741e348b9c6d86cc43ec5/Lib/os.py#L1067-L1078

import os
import abc
import sys
import typing as t
import pathlib
import unittest


class PathLike(abc.ABC):
    """Abstract base class for implementing the file system path protocol."""

    @abc.abstractmethod
    def __fspath__(self):
        """Return the file system path representation of the object."""
        raise NotImplementedError

    @classmethod
    def __subclasshook__(cls, subclass):
        return hasattr(subclass, '__fspath__')


# We monkey patch os here. We could probably avoid this, but this way we can
# copy paste the PathLike implementation + tests from StdLib.
if not hasattr(os, "PathLike"):
    os.PathLike = PathLike


class _PathLike(os.PathLike):
    """ Class implementing the file system path protocol."""

    def __init__(self, path=""):
        self.path = path

    def __str__(self):
        return str(self.path)

    def __fspath__(self):
        if isinstance(self.path, BaseException):
            raise self.path
        else:
            return self.path


def fspath(path) -> t.Union[str, bytes]:
    """Return the path representation of a path-like object.

    Returns
    -------
    - If str or bytes is passed in, it is returned unchanged.
    - If the os.PathLike interface is implemented it is used to get the path
      representation.
    - If the python version is 3.5 or earlier and a pathlib object is passed,
      the object's string representation is returned.

    Raises
    ------
    - Regardless of the input, if the path representation (e.g. the value
      returned from __fspath__) is not str or bytes, TypeError is raised.
    - If the provided path is not str, bytes, pathlib.PurePath or os.PathLike,
      TypeError is raised.
    """
    if isinstance(path, (str, bytes)):
        return path
    # Work from the object's type to match method resolution of other magic
    # methods.
    path_type = type(path)
    try:
        path_repr = path_type.__fspath__(path)
    except AttributeError:
        if hasattr(path_type, '__fspath__'):
            raise
        elif isinstance(path, pathlib.PurePath):
            # On python 3.5 pathlib objects don't have an __fspath__ method.
            # but we still want to get their string representation.
            return str(path)
        else:
            raise TypeError(
                "expected str, bytes or os.PathLike object, "
                "not " + path_type.__name__
            )
    if isinstance(path_repr, (str, bytes)):
        return path_repr
    else:
        raise TypeError(
            "expected {}.__fspath__() to return str or bytes, "
            "not {}".format(path_type.__name__,
                            type(path_repr).__name__)
        )


class TestPEP519(unittest.TestCase):

    fspath = staticmethod(fspath)

    def test_return_bytes(self):
        for b in b'hello', b'goodbye', b'some/path/and/file':
            self.assertEqual(b, self.fspath(b))

    def test_return_string(self):
        for s in 'hello', 'goodbye', 'some/path/and/file':
            self.assertEqual(s, self.fspath(s))

    @unittest.skipIf(
        sys.version_info < (3, 6),
        "fs{en,de}code() functions don't support os.PathLike on Python < 3.6"
    )
    def test_fsencode_fsdecode(self):
        for p in "path/like/object", b"path/like/object":
            pathlike = _PathLike(p)

            self.assertEqual(p, self.fspath(pathlike))
            self.assertEqual(b"path/like/object", os.fsencode(pathlike))
            self.assertEqual("path/like/object", os.fsdecode(pathlike))

    def test_pathlike(self):
        self.assertEqual('#feelthegil', self.fspath(_PathLike('#feelthegil')))
        self.assertTrue(issubclass(_PathLike, os.PathLike))
        self.assertTrue(isinstance(_PathLike(), os.PathLike))

    def test_pathlib_35(self):
        # pathlib on python 3.5 does not implement the PathLike protocol
        # so we need explicitly tests for pathlib.
        self.assertEqual(
            "/home/user", self.fspath(pathlib.PurePath("/home/user"))
        )
        self.assertEqual(
            "C:\\windows\\", self.fspath(pathlib.PurePath("C:\\windows\\"))
        )

    def test_garbage_in_exception_out(self):
        vapor = type('blah', (), {})
        for o in int, type, os, vapor():
            self.assertRaises(TypeError, self.fspath, o)

    def test_argument_required(self):
        self.assertRaises(TypeError, self.fspath)

    def test_bad_pathlike(self):
        # __fspath__ returns a value other than str or bytes.
        self.assertRaises(TypeError, self.fspath, _PathLike(42))
        # __fspath__ attribute that is not callable.
        c = type('foo', (), {})
        c.__fspath__ = 1
        self.assertRaises(TypeError, self.fspath, c())
        # __fspath__ raises an exception.
        self.assertRaises(
            ZeroDivisionError, self.fspath, _PathLike(ZeroDivisionError())
        )
```

**Comment by njsmith** [comment]
I was thinking it would be simplest to just write our own test that exercises the different paths here, especially the red ones: https://codecov.io/gh/python-trio/trio/src/master/trio/_util.py

I guess a full-fledged backport package has some advantages in theory, but it seems like a lot more work?

**Pull request merged**